# benchWGBSmap

Performance of mapping approaches for whole-genome bisulfite sequencing data in crop plants

Claudius Grehl1,2*, Marc Wagner3, Ioana Lemnian1,4, Bruno Glaser2, Ivo Grosse1,5

1Institute of Computer Science, Bioinformatics, Martin Luther University Halle–Wittenberg, Von Seckendorff-Platz 1, D-06120, Halle (Saale), Germany;
2Institute of Agronomy and Nutritional Sciences, Soil Biogeochemistry, Martin Luther University Halle–Wittenberg, Von–Seckendorff–Platz 3, D-06120, Halle (Saale), Germany;
3Freie Universität Berlin, Kaiserswerther Str. 16-18, D-14195, Berlin, Germany;
4Institute of Human Genetics, Martin Luther University Halle-Wittenberg, Magdeburger Str. 2, D-06112, Halle (Saale), Germany;
5German Centre for Integrative Biodiversity Research (iDiv) Halle-Jena-Leipzig, Deutscher Platz 5e, D-04103, Leipzig, Germany;

* Correspondence: 
Claudius Grehl
claudius.grehl@informatik.uni-halle.de

Keywords: Epigenetics, DNA methylation patterns, Read mapping, Benchmarking, WGBS


This is the README.txt file for the scripts of the manuscript "Performance of mapping approaches for whole-genome bisulfite sequencing data in crop plants".

